book/src/counter.md

```diff
@@ -0,0 +1,331 @@
+    // The pointer to the counter in the native (Rust's) heap
+    private long ptr;
+
+    // Don't worry if you came up with different names for the native methods
+    public Counter() {
+        this.ptr = NativeAPI.counter_new();
+    }
+
+    public int get() {
+        return NativeAPI.counter_get(ptr);
+    }
+
+    public int increment() {
+        return NativeAPI.counter_increment(ptr);
+    }
+
+    public void close() {
+        NativeAPI.counter_destroy(ptr);
+    }
+}
+```
+
+## Exercises
+
+1. Does garbage collecting a `Counter` free any native memory allocated by
+   `NativeAPI.new_counter`?
+2. What function signatures does the Rust library need for
+   `NativeAPI.counter_destroy(long)`, `NativeAPI.counter_get(long)`, and
+   `NativeAPI.counter_increment(long)` to work?
+
+## Rust Side
+
+Now that the Java interface is defined, it's time to implement the Rust side. There are two steps:
+
+1. Implement the `Counter` in normal Rust.
+2. Introduce the JNI functions that Java uses to interact with the Counter.
+
+### Implement the Counter
+
+We'll implement the counter first using normal Rust.
+
+```rust
+struct Counter {
+    count: i32,
+}
+
+impl Counter {
+    fn new() -> Self {
+        Counter {
+            count: 0,
+        }
+    }
+
+    fn increment(&mut self) -> i32 {
+        self.count += 1;
+        self.count
+    }
+
+    fn get(&self) -> i32 {
+        self.count
+    }
+}
+```
+
+### Introducing the JNI functions for the Counter
+
+Recall that we picked these APIs on the Java side:
+
+```
+static native long counter_new();
+static native int counter_get(long ptr);
+static native int counter_increment(long ptr);
+static native void counter_destroy(long ptr);
+```
+
+So, let's start by defining the native signatures: 
+
+```rust
+use jni::objects::JClass;
+use jni::sys::{jint, jlong};
+use jni::JNIEnv;
+
+#[no_mangle]
+pub extern "system" fn Java_com_github_jni_1rs_jnibook_NativeAPI_counter_1new(
+    _env: JNIEnv,
+    _class: JClass
+) -> jlong {
+    unimplemented!();
+}
+
+#[no_mangle]
+pub extern "system" fn Java_com_github_jni_1rs_jnibook_NativeAPI_counter_1get(
+    _env: JNIEnv,
+    _class: JClass,
+    ptr: jlong) -> jint {
+    unimplemented!();
+}
+
+#[no_mangle]
+pub extern "system" fn Java_com_github_jni_1rs_jnibook_NativeAPI_counter_1increment(
+    _env: JNIEnv,
+    _class: JClass,
+    ptr: jlong) -> jint {
+    unimplemented!();
+}
+
+#[no_mangle]
+pub extern "system" fn Java_com_github_jni_1rs_jnibook_NativeAPI_counter_1destroy(
+    _env: JNIEnv,
+    _class: JClass,
+    ptr: jlong) {
+    unimplemented!();
+}
+```
+
+Now, we have a few goals for the implementation.
+
+1. `counter_1new` should allocate a counter on the heap, and return a valid
+   pointer to it.
+2. `counter_1new`, `counter_1increment`, and `counter_1get` should not deallocate the counter.
+3. `counter_1destroy` should deallocate the counter.
+
+We'll need to use [`Box`](https://doc.rust-lang.org/std/boxed/index.html) to
```

Comment:
I wish we already had https://github.com/jni-rs/jni-rs/issues/84 moved from EJB into a satellite library to not have to discuss this :)

Reply:
Yeah that does look helpful. Fortunately, it's not a one-way door. Once #84 is resolved, we can decide how to introduce it here.